idx/memory/memory.go

```diff
@@ -779,66 +779,73 @@ func (m *UnpartitionedMemoryIdx) GetPath(orgId uint32, path string) []idx.Archiv
 	return archives
 }
 
-func (m *UnpartitionedMemoryIdx) TagDetails(orgId uint32, key string, filter *regexp.Regexp, from int64) map[string]uint64 {
+func (m *UnpartitionedMemoryIdx) FindByTag(orgId uint32, query tagquery.Query) []idx.Node {
 	if !TagSupport {
 		log.Warn("memory-idx: received tag query, but tag support is disabled")
 		return nil
 	}
 
+	queryCtx := NewTagQueryContext(query)
+
 	m.RLock()
 	defer m.RUnlock()
 
-	tags, ok := m.tags[orgId]
+	tagIndex, ok := m.tags[orgId]
 	if !ok {
+		// if there is no tag index, we don't need to run the tag query
 		return nil
 	}
 
-	values, ok := tags[key]
-	if !ok {
-		return nil
+	var enricher *enricher
+	if metaTagSupport {
+		mtr, ok := m.metaTagRecords[orgId]
+		if ok {
+			enricher = mtr.getEnricher(tagIndex.idHasTag)
+		}
 	}
 
-	res := make(map[string]uint64)
-	for value, ids := range values {
-		if filter != nil && !filter.MatchString(value) {
+	// construct the output slice of idx.Node's such that there is only 1 idx.Node for each path
+	resCh := m.idsByTagQuery(orgId, queryCtx)
+
+	byPath := make(map[string]*idx.Node)
+	for id := range resCh {
+		def, ok := m.defById[id]
+		if !ok {
+			corruptIndex.Inc()
+			log.Errorf("memory-idx: corrupt. ID %q has been given, but it is not in the byId lookup table", id)
 			continue
 		}
 
-		count := uint64(0)
-		if from > 0 {
-			for id := range ids {
-				def, ok := m.defById[id]
-				if !ok {
-					corruptIndex.Inc()
-					log.Errorf("memory-idx: corrupt. ID %q is in tag index but not in the byId lookup table", id)
-					continue
-				}
-
-				if atomic.LoadInt64(&def.LastUpdate) < from {
-					continue
-				}
-
-				count++
+		nameWithTags := def.NameWithTags()
+		if existing, ok := byPath[nameWithTags]; !ok {
+			byPath[nameWithTags] = &idx.Node{
+				Path:        nameWithTags,
+				Leaf:        true,
+				HasChildren: false,
+				Defs:        []idx.Archive{CloneArchive(def)},
+			}
+			if enricher != nil {
+				byPath[nameWithTags].MetaTags = enricher.enrich(def.Id, def.Name, def.Tags)
 			}
 		} else {
-			count += uint64(len(ids))
+			existing.Defs = append(existing.Defs, CloneArchive(def))
 		}
+	}
 
-		if count > 0 {
-			res[value] = count
-		}
+	results := make([]idx.Node, len(byPath))
+
+	i := 0
+	for _, v := range byPath {
+		results[i] = *v
```

Comment:
I like the idea of changing it in the future.

I didn't mean that we should return `map[string]idx.Node`, I just meant we should use it internally in that method instead of `map[string]*idx.Node`. We are creating a map of pointers and then immediately dereferencing it when we add it to the slice.

If we decide later to return a `[]*idx.Node` then changing it now doesn't make sense either I suppose.

Reply:
Right, I'm talking about the current state of things. We are not returning `[]*idx.Node`, so why are we creating a map of pointers and then just converting them straight away?